Source: Elias Pipping <pipping@exherbo.org>
Upstream: Attached to http://llvm.org/bugs/show_bug.cgi?id=8579
Reason: Linking basically anything with clang would fail

From c73307882ad8c01e4b58c8dd3d1a2db1df30706a Mon Sep 17 00:00:00 2001
From: Elias Pipping <pipping@exherbo.org>
Date: Wed, 10 Nov 2010 13:49:28 +0100
Subject: [PATCH] Add support for Exherbo Linux

---
 lib/Driver/ToolChains.cpp |   13 ++++++++++++-
 1 files changed, 12 insertions(+), 1 deletions(-)

diff --git a/lib/Driver/ToolChains.cpp b/lib/Driver/ToolChains.cpp
index 680a17f..bd67572 100644
--- a/lib/Driver/ToolChains.cpp
+++ b/lib/Driver/ToolChains.cpp
@@ -1198,6 +1198,7 @@ Tool &AuroraUX::SelectTool(const Compilation &C, const JobAction &JA) const {
 enum LinuxDistro {
   DebianLenny,
   DebianSqueeze,
+  Exherbo,
   Fedora13,
   Fedora14,
   OpenSuse11_3,
@@ -1227,8 +1228,12 @@ static bool IsDebianBased(enum LinuxDistro Distro) {
 }
 
 static bool HasMultilib(llvm::Triple::ArchType Arch, enum LinuxDistro Distro) {
-  if (Arch == llvm::Triple::x86_64)
+  if (Arch == llvm::Triple::x86_64) {
+    if (Distro == Exherbo && ! llvm::sys::Path("/usr/lib32/libc.so").exists())
+      return false;
+
     return true;
+  }
   if (Arch == llvm::Triple::x86 && IsDebianBased(Distro))
     return true;
   return false;
@@ -1281,6 +1286,9 @@ static LinuxDistro DetectLinuxDistro(llvm::Triple::ArchType Arch) {
     return UnknownDistro;
   }
 
+  if (llvm::sys::Path("/etc/exherbo-release").exists())
+    return Exherbo;
+
   return UnknownDistro;
 }
 
@@ -1314,6 +1322,9 @@ Linux::Linux(const HostInfo &Host, const llvm::Triple& Triple)
   } else if (Arch == llvm::Triple::x86_64) {
     if (llvm::sys::Path("/usr/lib/gcc/x86_64-linux-gnu").exists())
       GccTriple = "x86_64-linux-gnu";
+    else if (llvm::sys::Path("/usr/lib/gcc/x86_64-pc-linux-gnu").exists())
+      // This is actually a quadruple
+      GccTriple = "x86_64-pc-linux-gnu";
     else if (llvm::sys::Path("/usr/lib/gcc/x86_64-redhat-linux").exists())
       GccTriple = "x86_64-redhat-linux";
     else if (llvm::sys::Path("/usr/lib64/gcc/x86_64-suse-linux").exists())
-- 
1.7.3.2

